src/core/value/constant.jl

```diff
@@ -102,53 +102,175 @@ export ConstantAggregateZero
 end
 identify(::Type{Value}, ::Val{API.LLVMConstantAggregateZeroValueKind}) = ConstantAggregateZero
 
-# there currently seems to be no function in the LLVM-C interface which returns a
-# ConstantAggregateZero value directly, but values can occur through calls to LLVMConstNull
+# array interface
+# FIXME: can we reuse the ::ConstantArray functionality with ConstantAggregateZero values?
+#        probably works fine if we just get rid of the refcheck
+Base.eltype(caz::ConstantAggregateZero) = eltype(llvmtype(caz))
+Base.size(caz::ConstantAggregateZero) = (0,)
+Base.length(caz::ConstantAggregateZero) = 0
+Base.axes(caz::ConstantAggregateZero) = (Base.OneTo(0),)
+Base.collect(caz::ConstantAggregateZero) = Value[]
 
 
-## constant expressions
-
-export ConstantExpr, ConstantAggregate, ConstantArray, ConstantStruct, ConstantVector, InlineAsm
-
-@checked struct ConstantExpr <: Constant
-    ref::API.LLVMValueRef
-end
-identify(::Type{Value}, ::Val{API.LLVMConstantExprValueKind}) = ConstantExpr
+## regular aggregate
 
 abstract type ConstantAggregate <: Constant end
 
+# arrays
+
 @checked struct ConstantArray <: ConstantAggregate
     ref::API.LLVMValueRef
 end
 identify(::Type{Value}, ::Val{API.LLVMConstantArrayValueKind}) = ConstantArray
 identify(::Type{Value}, ::Val{API.LLVMConstantDataArrayValueKind}) = ConstantArray
 
-ConstantArray(typ::LLVMType, data::Vector{T}) where {T<:Constant} =
-    ConstantArray(API.LLVMConstArray(typ, data, length(data)))
-ConstantArray(typ::IntegerType, data::Vector{T}) where {T<:Integer} =
-    ConstantArray(typ, map(x->ConstantInt(convert(T,x),context(typ)), data))
-ConstantArray(typ::FloatingPointType, data::Vector{T}) where {T<:AbstractFloat} =
-    ConstantArray(typ, map(x->ConstantFP(convert(T,x),context(typ)), data))
+ConstantArrayOrAggregateZero(value) = Value(value)::Union{ConstantArray,ConstantAggregateZero}
 
-Base.getindex(ca::ConstantArray, idx::Integer) =
-    API.LLVMGetElementAsConstant(ca, idx-1)
-Base.length(ca::ConstantArray) = length(llvmtype(ca))
+# generic constructor taking an array of constants
+function ConstantArray(data::AbstractArray{<:Constant,N},
+                       typ::LLVMType=llvmtype(first(data))) where {N}
+    @assert all(x->x==typ, llvmtype.(data))
+
+    if N == 1
+        return ConstantArrayOrAggregateZero(API.LLVMConstArray(typ, Array(data), length(data)))
+    end
+
+    if VERSION >= v"1.1"
+        ca_vec = map(x->ConstantArray(x, typ), eachslice(data, dims=1))
+    else
+        ca_vec = map(x->ConstantArray(x, typ), (view(data, i, ntuple(d->(:), N-1)...) for i in axes(data, 1)))
+    end
+    ca_typ = llvmtype(first(ca_vec))
+
+    return ConstantArray(API.LLVMConstArray(ca_typ, ca_vec, length(ca_vec)))
+end
+
+# shorthands with arrays of plain Julia data
+ConstantArray(data::AbstractArray{T,N}, ctx::Context=GlobalContext()) where {T<:Integer,N} =
```

Comment:
This doesn't work when `T` is `Bool` (`<:Integer`). Maybe we should add a specific method to `ConstantInt` for `Bool` that creates a constant with type `Int1Type`.

Reply:
Not really related to this PR, but OK, I pushed a fix.